services/src/main/java/org/keycloak/services/resources/admin/AdminConsole.java

```diff
@@ -213,38 +221,32 @@ public Response whoAmI(final @Context HttpHeaders headers) {
             displayName = user.getUsername();
         }
 
-        RealmModel masterRealm = getAdminstrationRealm(realmManager);
-        Map<String, Set<String>> realmAccess = new HashMap<String, Set<String>>();
-        if (masterRealm == null)
+        RealmModel currentRealm = session.realms().getRealmByName(realmName);
+
+        if (currentRealm == null)
             throw new NotFoundException("No realm found");
-        boolean createRealm = false;
-        if (realm.equals(masterRealm)) {
-            logger.debug("setting up realm access for a master realm user");
-            createRealm = user.hasRole(masterRealm.getRole(AdminRoles.CREATE_REALM));
-            addMasterRealmAccess(user, realmAccess);
-        } else {
-            logger.debug("setting up realm access for a realm user");
-            addRealmAccess(realm, user, realmAccess);
-        }
 
+        Map<String, Set<String>> roles = calculateRoles(currentRealm, user);
+        RealmModel masterRealm = new RealmManager(session).getKeycloakAdminstrationRealm();
+        boolean createRealm = user.hasRole(masterRealm.getRole(AdminRoles.CREATE_REALM));
         Locale locale = session.getContext().resolveLocale(user);
 
-        return Response.ok(new WhoAmI(user.getId(), realm.getName(), displayName, createRealm, realmAccess, locale)).build();
+        return Response.ok(new WhoAmI(user.getId(), realm.getName(), displayName, createRealm, roles, locale)).build();
     }
 
-    private void addRealmAccess(RealmModel realm, UserModel user, Map<String, Set<String>> realmAdminAccess) {
-        RealmManager realmManager = new RealmManager(session);
-        ClientModel realmAdminApp = realm.getClientByClientId(realmManager.getRealmAdminClientId(realm));
-        getRealmAdminAccess(realm, realmAdminApp, user, realmAdminAccess);
-    }
+    private Map<String, Set<String>> calculateRoles(RealmModel realm, UserModel user) {
+        ClientModel client = realm.getMasterAdminClient();
+        Set<String> realmRoles = client.getRolesStream()
+                .filter(user::hasRole)
+                .map(RoleModel::getName)
+                .collect(Collectors.toSet());
 
-    private void addMasterRealmAccess(UserModel user, Map<String, Set<String>> realmAdminAccess) {
-        session.realms().getRealmsStream().forEach(realm -> {
-            ClientModel realmAdminApp = realm.getMasterAdminClient();
-            getRealmAdminAccess(realm, realmAdminApp, user, realmAdminAccess);
-        });
-    }
+        Map<String, Set<String>> realmAccess = new HashMap<>(20);
+
+        realmAccess.put(realm.getName(), realmRoles);
 
+        return realmAccess;
+    }
```

Comment:
Can this be simplified to?
```suggestion
        return Collections.singletonMap(realm.getName(), realmRoles);
    }
```

Reply:
Done